src/tools/cgexec.c

```diff
@@ -171,12 +188,128 @@ int main(int argc, char *argv[])
+	 *    a task with /proc/<pid>/cmdline as
+	 *    "libcgroup_systemd_idle_thread", kill it.  This gives the
+	 *    illusion of replacement.
+	 */
+	exec_task_pid = fork();
+	if (exec_task_pid == -1) {
+		err("Fork failed:%s\n", strerror(errno));
+		return  -1;
+	}
+
+	/* child process exec's into task program */
+	if (exec_task_pid == 0) {
+		execvp(argv[optind], &argv[optind]);
+		err("exec failed:%s", strerror(errno));
+		return  -1;
+	}
+
+	snprintf(pid_proc_cgrp_path, FILENAME_MAX, "/proc/%u/cgroup", pid);
+	pid_proc_cgrp_fp = fopen(pid_proc_cgrp_path, "re");
+	if (!pid_proc_cgrp_fp) {
+		err("Failed to open: %s\n", pid_proc_cgrp_path);
 		return -1;
 	}
 
+	while (fgets(buffer, FILENAME_MAX, pid_proc_cgrp_fp)) {
+		char ctrl_name[CONTROL_NAMELEN_MAX];
+		char *ctrl;
+		int len;
+
+		/* read according to the cgroup mode */
+		if (strstr(buffer, "::"))
+			ret = sscanf(buffer, "%d::%4096s\n", &id, cgroup_name);
+		else
+			ret = sscanf(buffer, "%d:%[^:]:%4096s\n", &id, ctrl_name, cgroup_name);
+
+		/* skip named controllers */
+		if (ret == 3 && strncmp(ctrl_name, "name", 4) == 0)
+			continue;
+
+		/* skip if the cgroup doesn't ends with .scope */
+		base = basename(cgroup_name);
+		len = strlen(base) - 6;
+		if (strcmp(base + len, ".scope") != 0)
+			continue;
+
+		/* cgroup v2 (unified) mode */
+		if (ret == 2) {
+			scope_pid = find_scope_idle_task_pid(cgroup_name, NULL);
+		} else {
+			/* cgroup v1 (legacy/hybrid) mode */
+			/* check if the mount point is shared between controllers */
+			ctrl = strchr(ctrl_name, ',');
+			if (ctrl) {
+				/* truncate other controllers, but first */
+				len = strlen(ctrl_name) - strlen(ctrl);
+				ctrl_name[len] = '\0';
+			}
+			scope_pid = find_scope_idle_task_pid(cgroup_name, ctrl_name);
+		}
+
+		if (scope_pid > 1) {
+			/* kill the idle thread process */
+			ret = kill(scope_pid, SIGTERM);
+			if (ret) {
+				/*
+				 * Both scope idle_thread and new task
+				 * will be visible under scope
+				 */
+				err("Failed to replace scope idle_thread pid %u:%s\n",
+				    scope_pid, strerror(errno));
+				fclose(pid_proc_cgrp_fp);
+				return  -1;
+			}
+		}
+	}
+	fclose(pid_proc_cgrp_fp);
+
 	return 0;
 }
+
+static int find_scope_idle_task_pid(char *cgroup_name, char *ctrl_name)
+{
+	char pid_proc_cmd_path[FILENAME_MAX], buffer[FILENAME_MAX];
+	FILE *pid_proc_cmd_fp = NULL;
+	int i, ret, size;
+	pid_t *pids;
+
+	ret = cgroup_get_procs(cgroup_name, ctrl_name, &pids, &size);
+	if (ret) {
+		err("Failed to read cgroup.procs of cgroup: %s\n", cgroup_name + 1);
+		return -1;
+	}
+
+	for(i = 0; i < size; i++) {
+		snprintf(pid_proc_cmd_path, FILENAME_MAX, "/proc/%u/cmdline", pids[i]);
+		pid_proc_cmd_fp = fopen(pid_proc_cmd_path, "re");
+		/* task might have exited */
+		if (!pid_proc_cmd_fp)
+			continue;
+
+		/* task might have exited, so consider only successful reads. */
+		if (fgets(buffer, FILENAME_MAX, pid_proc_cmd_fp)) {
+			if (strcmp(buffer, "libcgroup_systemd_idle_thread") == 0) {
```

Comment:
Should we make `libcgroup_systemd_idle_thread` a constant?

Reply:
Good idea, sure.